src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -0,0 +1,500 @@
+        /// <remarks>
+        /// This method override is for registering an aggregated COM instance with its associated inner. The inner
+        /// will be released when the associated wrapper is eventually freed. Note that it will be released on a thread
+        /// in an unknown apartment state. If the supplied inner is not known to be a free-threaded instance then
+        /// it is advised to not supply the inner.
+        ///
+        /// If the <paramref name="wrapper"/> instance already has an associated external object a <see cref="System.NotSupportedException"/> will be thrown.
+        /// </remarks>
+        public object GetOrRegisterObjectForComInstance(IntPtr externalComObject, CreateObjectFlags flags, object wrapper, IntPtr inner)
+        {
+            if (wrapper == null)
+                throw new ArgumentNullException(nameof(wrapper));
+
+            object? obj;
+            if (!TryGetOrCreateObjectForComInstanceInternal(this, externalComObject, inner, flags, wrapper, out obj))
+                throw new ArgumentNullException(nameof(externalComObject));
+
+            return obj!;
+        }
+
+        /// <summary>
+        /// Get the currently registered managed object or creates a new managed object and registers it.
+        /// </summary>
+        /// <param name="impl">The <see cref="ComWrappers" /> implementation to use when creating the managed object.</param>
+        /// <param name="externalComObject">Object to import for usage into the .NET runtime.</param>
+        /// <param name="innerMaybe">The inner instance if aggregation is involved</param>
+        /// <param name="flags">Flags used to describe the external object.</param>
+        /// <param name="wrapperMaybe">The <see cref="object"/> to be used as the wrapper for the external object.</param>
+        /// <param name="retValue">The managed object associated with the supplied external COM object or <c>null</c> if it could not be created.</param>
+        /// <returns>Returns <c>true</c> if a managed object could be retrieved/created, <c>false</c> otherwise</returns>
+        /// <remarks>
+        /// If <paramref name="impl" /> is <c>null</c>, the global instance (if registered) will be used.
+        /// </remarks>
+        private static bool TryGetOrCreateObjectForComInstanceInternal(
+            ComWrappers impl,
+            IntPtr externalComObject,
+            IntPtr innerMaybe,
+            CreateObjectFlags flags,
+            object? wrapperMaybe,
+            out object? retValue)
+        {
+            if (externalComObject == IntPtr.Zero)
+                throw new ArgumentNullException(nameof(externalComObject));
+
+            if (flags.HasFlag(CreateObjectFlags.Aggregation))
+                throw new NotImplementedException();
+
+            object? wrapperMaybeLocal = wrapperMaybe;
+            retValue = impl.CreateObject(externalComObject, flags);
```

Comment:
This will always create a new object. That does not sound right. I think it shoiuld instead find out whether we have a existing object for the passed in IUnknown and just return that.

Reply:
if `CreateObjectFlags.Unique` passed that seems to be responsibility of ComWrapper implementation to make sure that this object is unique, and as such when Unique do not passed then caching is responsibility of ComWrappers too. If not can you clarify why `CreateObjectFlags.Unique` needed? I will remove RCW code from this PR, just since this all started, I want to understand problem better.

Reply:
It is to support https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getuniqueobjectforiunknown?view=net-5.0 .

It is useful when the shared cached wrapper is owned by somebody else and you want a different lifetime control for your instance.

Reply:
@jkotas Who maintain uniqueness of the `object` in that case, runtime, or `ComWrappers`? My understanding that `ComWrappers` maintain uniqueness of the objects based on passed flags, and runtime just maintain mapping between COM and proxy objects.

Otherwise if runtime maintain uniqueness of the object, what's the point of passing that specific flag to `ComWrappers`. If uniqueness would be guaranteed by runtime, then ComWrappers always can create unique instances and do not worry to much about that aspect.

Reply:
I agree that `CreateObject` is unlikely to check the `UniqueInstance` flag. I think that the flags are passed through unmodified from `GetOrCreateObjectForComInstance` to `CreateObject` for simplicity and future proofing.